src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,353 @@
+@file:Suppress("MemberVisibilityCanBePrivate")
+
+package app.revanced.patcher.apk
+
+import app.revanced.patcher.DomFileEditor
+import app.revanced.patcher.Patcher
+import app.revanced.patcher.PatcherOptions
+import app.revanced.patcher.arsc.*
+import app.revanced.patcher.util.ProxyBackedClassList
+import app.revanced.patcher.util.xml.LazyXMLInputSource
+import app.revanced.patcher.util.xml.scanIdRegistrations
+import com.reandroid.apk.ApkModule
+import com.reandroid.apk.xmlencoder.EncodeException
+import com.reandroid.apk.xmlencoder.EncodeMaterials
+import com.reandroid.apk.xmlencoder.EncodeUtil
+import com.reandroid.archive.InputSource
+import com.reandroid.arsc.chunk.PackageBlock
+import com.reandroid.arsc.chunk.TableBlock
+import com.reandroid.arsc.chunk.xml.AndroidManifestBlock
+import com.reandroid.arsc.value.Entry
+import com.reandroid.arsc.value.ResConfig
+import lanchon.multidexlib2.BasicDexEntry
+import lanchon.multidexlib2.DexIO
+import lanchon.multidexlib2.MultiDexContainerBackedDexFile
+import lanchon.multidexlib2.MultiDexIO
+import lanchon.multidexlib2.RawDexIO
+import org.jf.dexlib2.dexbacked.DexBackedDexFile
+import org.jf.dexlib2.iface.MultiDexContainer
+import org.jf.dexlib2.writer.io.MemoryDataStore
+import org.w3c.dom.*
+import java.io.File
+import java.util.zip.ZipEntry
+
+
+/**
+ * The [Apk] file that is to be patched.
+ *
+ * @param path The path to the apk file.
+ * @param name The name of this apk.
+ */
+sealed class Apk private constructor(internal val module: ApkModule) {
+    companion object {
+        const val manifest = "AndroidManifest.xml"
+    }
+
+    /**
+     * The metadata of the [Apk].
+     */
+    val packageMetadata = PackageMetadata(module.androidManifestBlock)
+
+    internal val archive = Archive(module)
+
+
+    /**
+     * Refresh updated resources for an [Apk].
+     *
+     * @param options The [PatcherOptions] of the [Patcher].
+     */
+    internal open fun finalize(options: PatcherOptions) {
+        archive.openFiles.forEach { options.logger.warn("File $it was never closed! File modifications will not be applied if you do not close them.") }
```

Comment:
Shouldn't they be closed here?

Reply:
Well, the `openFiles` set is just used to lock files. We could make the patcher close open files, but not closing files seems like a developer error anyway, so do we really need to add that feature?

Reply:
I mean, it mitigates developer error. The developer could forget to close it but it would be fine since it closes anyways. If it doesn't close, the chance that something fails would be higher.

Reply:
Sure